Introduce #paranoia_destroy_attributes and #paranoia_restore_attributes extension points

Empact opened the pull request:
We maintain a second `active` column for the sake of enforcing uniqueness among un-deleted rows - this is a sort of more general solution to the same problem https://github.com/radar/paranoia/wiki/Custom-sentinel-values attempts to solve (incidentally, a problem more elegantly solved by http://www.postgresql.org/docs/9.4/static/indexes-expressional.html, if one has access to them).

As an alternative to maintaining this extra value via a separate callback, it would be nice to update it at the same time paranoia updates the `deleted_at` column. We can do that and perhaps serve other needs by introducing these 2 extension points. With them, getting the behavior we need is a simple matter of adding a value to each of the hashes from these methods.

One significant difference brought on by this change is that `#touch` occurs inside a transaction, whereas `#update_columns` creates no such transaction. This only affects `#delete` because `#destroy` already creates its own transaction. To my eye, this behavior is more in line with the activerecord documentation and default behavior, which makes no mention of transactions for `#delete` but it is a change.

Empact committed:
Inline touch_paranoia_column into delete

Empact committed:
Add paranoia_destroy_attributes and paranoia_restore_attributes as extension points

Use update_columns rather than touch to update the record, for
generality

Unlike touch, update_columns does not create a transaction for
itself, so we need to add the record to the transaction, if present.
If there is not a current_transaction, the add is a no-op.

This all means that delete will not invoke a transaction or run the
after_commit callbacks unless called from within one, which is
consistent with the Rails docs and the behavior of
ActiveRecord::Base#delete.

Empact commented:
The alternative we're taking advantage of at the moment is to override the behavior of `touch_paranoia_column` and `restore!`, which is more duplicative than I would like.

Empact commented:
Failure looks spurious to me.

radar commented:
Looks good to me. Could you please add some documentation for this to the README?

Empact commented:
I updated the branch here: https://github.com/brigade/paranoia/tree/update_attributes doesn't seem to be showing up in the diff here. Changes are that I update the readme, add some testing, and fixed an issue in .only_deleted, which would miss records when a null value signifies deletion.

Empact commented:
@radar I think this is ready to go

radar commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 # paranoia Changelog
 
+## Unreleased
```
I've done a bad job of keeping this up to date. Normally what I would do is just point people at the changes between tags.

Empact replied:
Would you rather I drop it then?

radar replied:
Yup, that'd be preferable.

Empact replied:
:+1:

radar commented on README.md:
```diff
@@ -207,6 +207,49 @@ add_index :clients, [:group_id, :other_id], where: "deleted_at IS NULL"
+on the deleted_at column and expect it to enforce that there only be one record
+with a certain combination of values.
+
+If your database supports g, good alternatives include partial indexes
```
"supports g" <- is that right?

radar commented on lib/paranoia.rb:
```diff
@@ -29,7 +29,16 @@ def with_deleted
 
     def only_deleted
-      with_deleted.where.not(paranoia_column => paranoia_sentinel_value)
+      if paranoia_sentinel_value == nil
```
Prefer `paranoia_sentinel_value.nil?` here.

radar commented:
LGTM aside from the comments I've left.

radar commented:
This PR does not merge cleanly to master. Please update.

Empact commented:
@radar Done

radar commented:
Thanks!
